eox_nelp/programs/api/v1/views.py

```diff
@@ -106,3 +144,56 @@ def post(self, request, course_id):
         update_program_metadata(course_id, serializer.data, request.user)
 
         return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+
+class ProgramsListView(CourseListView):
+    """
+    API view to list courses with program metadata.
+    **Use Cases**
+        - List visible courses with associated program metadata for request user
+        - List courses in which the user identified by national_id is enrolled,
+          provided the requesting user has permission to view them.
+
+    **Example Requests**
+        GET /eox-nelp/api/programs/v1/program-lookup/
+        GET /eox-nelp/api/programs/v1/program-lookup/?national_id=1234567890
+    """
+    authentication_classes = [JwtAuthentication, SessionAuthenticationAllowInactiveUser]
+    permission_classes = [IsAuthenticated, ProgramsLookupPermission]
+    serializer_class = ProgramLookupSerializer
+
+    @process_national_id_query_param()
+    def get(self, request, *args, **kwargs):
+        """Override get to apply some filtering and pre processing."""
+        return super().get(request, *args, **kwargs)
+
+    def get_queryset(self):
+        """Override qs to query by national_id if provided.
+        Returns:
+            QuerySet: queryset of courses depending on national_id presence.
+            If national_id is provided, returns courses the user nid.
+            If not provided, returns all courses visible to the request user.
+        """
+        if getattr(self.request, 'nid_user', None):
+            return courses.get_courses(user=self.request.nid_user)
+        return super().get_queryset()  # Visible courses for user queryset
+
+    def filter_queryset(self, queryset):
+        """
+        Filter the queryset by course enrolled if the national_id query param is present.
+        Returns:
+        List of enrolled courses for the user qs: Depending on performance this list queryset  could be improved to
+        lazy sequence as used in
+        https://github.com/openedx/edx-platform/blob/258f3fc/lms/djangoapps/course_api/api.py#L111
+        """
+        if getattr(self.request, 'nid_user', None):
+            queryset = [
+                course for course in queryset if CourseEnrollment.is_enrolled(self.request.nid_user, str(course.id))
+            ]
+        program_queryset = []
+        for course in queryset:
+            course_data = CourseDetailSerializer(course, context={'request': self.request}).data
+            program_lookup = get_program_lookup_representation(course_data)
+            program_queryset.append(program_lookup)
+
+        return program_queryset
```

Comment:
why did you choose to place this logic here ? if I want to use the `get_queryset` in another place won't be possible to get  the program metadata